### PR TITLE
Migrate user-facing apis to null-safety

### LINKED ISSUE #2920: Add support for nullsafety
I am upgrading my package (Kiwi) vanlooverenkoen/kiwi#51

But it seems that build & build_runner is not yet migrated to support nullsafety. Is there a timeline on when this could be added?

### PR DESCRIPTION
**Pull request by simolus3**
Migrate packages typically imported by build authors to null-safety (part 1 /3). Closes #2920

In this PR, I migrate

- the `build` package
- the `build_test` package (or rather, the parts of that package that don't import `build_resolvers`)
- the `scratch_space` package

After the `graphs` package is migrated, I'll open another PR to migrate the `build_resolvers` package. When that one is done too, we can start to opt the tests of `build` and `build_test` into null-safety and publish, thus enabling build authors to write and test their builders with strong null-safety. 
Migrating the rest of the build system is not a goal for the near future, so we won't run builds with strong null-safety.

### CODE REVIEW

File: build_test/lib/src/resolve_source.dart
Comment (jakemac53) [inline build_test/lib/src/resolve_source.dart:121]: Lets go ahead and make this non-nullable - it was already doing a null check right away :).

You can use `ArgumentError.checkNotNull` to retain the check for legacy codebase interactions, if desired (but I wouldn't bother)

File: build_test/lib/src/stub_reader.dart
Comment (jakemac53) [inline build_test/lib/src/stub_reader.dart:25]: I am confused how the other methods don't need to be updated?

File: build_test/lib/src/test_bootstrap_builder.dart
Comment (jakemac53) [inline build_test/lib/src/test_bootstrap_builder.dart:10]: We don't want to be ignoring any of these at least not under `lib` - we can push on migrating `test_core/backend.dart` sooner than later, but for now I would opt out this file and any file that imports it.

Otherwise when that library does migrate it could break this library.

File: build_test/pubspec.yaml
Comment (jakemac53) [inline build_test/pubspec.yaml:10]: We should up all deps to require the null safe versions (we are likely relying on them in practice)

```suggestion
  async: ^2.5.0
```

**Comment by jakemac53** [comment]
Thanks for starting on this! As you have probably discovered this is going to be a large and somewhat complicated effort. My thoughts were that we should be able to migrate mostly in a bottom up fashion, one package at a time. I do believe it is feasible to do that, although I am not sure all of the packages are unblocked yet, and we will want to wait on that probably before publishing _any_ of them (I think it just gets confusing otherwise, and it doesn't unblock anybody). 

Tests would initially all be opted out (or run in --no-sound-null-safety mode). Maybe the latter is better.

The order would be roughly [`build`, `build_config`, `build_daemon`], [`build_resolvers`, `scratch_space`], `build_test`, `build_runner_core`, `build_runner`, `build_modules`, [`build_web_compilers`, `build_vm_compilers`].

Then we would go back through and enable sound null safety on all the tests, fixing things up as needed.

I think probably we would want to do all of this before we publish any of them, get them all passing tests and happy, etc. That means a somewhat long feature freeze most likely. Whenever one is migrated you would have to go add overrides for it in all the ones that depend on it.

Another option we may want to consider, is splitting this into two chunks for publishing. If we get `build`, `build_config`, `build_resolvers`, and `build_test` migrated, that should be pretty much everything any builders depend on, and most builders would at that point be unblocked from migrating. I think we could run most tests with sound null safety at that point also. We could then publish those.

The rest of the packages are more related to the build script itself, which can be tackled as a separate thing, and is honestly just a non-goal for the immediate future.

**Comment by jakemac53** [comment]
cc @natebosch

**Comment by simolus3** [comment]
Thanks for the quick feedback!

> Another option we may want to consider, is splitting this into two chunks for publishing. If we get `build`, `build_config`, `build_resolvers`, and `build_test` migrated, that should be pretty much everything any builders depend on, and most builders would at that point be unblocked from migrating

From a builder-author perspective, this is what I was aiming for. I want to write and test my builders with sound null safety, but I don't care about whether the actual build runner is using null safety or not (similar to the story with `package:test`).
Maybe we can encourage users to run builds with weak null assertions if that helps.
And on a another thought we might not even need `build_config` for the first step: It's a non-dev dependency, but we typically don't import that into `lib/`.

`build`, `build_test` and `build_resolvers` have circular (dev-)dependencies, so we kind of have to migrate them at once. When that's done, we can run the majority of tests with sound null safety too.

File: build_test/lib/src/stub_reader.dart
Comment (simolus3) [reply build_test/lib/src/stub_reader.dart:25]: They all return `Future.value(null)` which apparently isn't a static error, but would fail at runtime.

`StubAssetReader`/`StubAssetWriter` seem to be exclusively used in other tests of the build system, for instance to ensure that a `BuildStepImpl` doesn't allow certain reads or writes. I wasn't sure how migrate the reader: Should the methods throw an `UnsupportedError('stub')`? Some bogus value?

File: build/test/generate/run_builder_test.dart
Comment (simolus3) [reply build/test/generate/run_builder_test.dart:1]: They were importing `package:build_test` before I migrated it. I think most of these can be removed as soon as the "core" packages are all migrated.

File: build/lib/src/generate/run_builder.dart
Comment (simolus3) [reply build/lib/src/generate/run_builder.dart:39]: The problem was that it's later used in the local `buildForInput` method, in which it wasn't promoted.

**Comment by jakemac53** [comment]
> `build`, `build_test` and `build_resolvers` have circular (dev-)dependencies, so we kind of have to migrate them at once. When that's done, we can run the majority of tests with sound null safety too.

They can be migrated separately - just the tests have to be opted out (or ran without sound null safety and add an ignore to the imports).

**Comment by jakemac53** [comment]
I am OK though with a larger PR that migrates them all at once, but its more difficult to manage and review etc. It is easier to have a few smaller PRs and then go back and enable null safety in a followup imo.

**Comment by simolus3** [comment]
Fair enough, I'll open another PR for `build_resolvers` and to migrate the tests later. I already migrated most of `build`, `build_test` and `scratch_space` though, if it's not too much work for you I'd prefer to not split those up. 

And unfortunately I had to add lots of dependency overrides now: Since `_test` wants all build packages from path, they all need to support the latest analyzer and glob, and due to breaking changes I couldn't really expand the version range. So the overrides are kind of starting to cascade now...

**Comment by jakemac53** [comment]
Ya I am fine with keeping this PR as is at this point and not splitting it up.

**Comment by jakemac53** [comment]
> And unfortunately I had to add lots of dependency overrides now: Since `_test` wants all build packages from path, they all need to support the latest analyzer and glob, and due to breaking changes I couldn't really expand the version range. So the overrides are kind of starting to cascade now...

Ya, this mono_repo is definitely more than a bit crazy like that

File: build_test/lib/src/stub_reader.dart
Comment (jakemac53) [reply build_test/lib/src/stub_reader.dart:25]: Ya we would have to look more at how the tests are used.... this type of pattern is definitely more challenging with null safety.

**Comment by simolus3** [comment]
I think we might want to put this on hold until the test packages support the latest analyzer. The dependency conflicts are getting out of hand.

**Comment by jakemac53** [comment]
> I think we might want to put this on hold until the test packages support the latest analyzer. The dependency conflicts are getting out of hand.

sgtm

**Comment by simolus3** [comment]
I'll wait for #3003 to land before fixing conflicts (otherwise I'd effectively have to do it twice). Apart from that, I think this should be ready now.

**Comment by jakemac53** [comment]
> I'll wait for #3003 to land before fixing conflicts (otherwise I'd effectively have to do it twice). Apart from that, I think this should be ready now.

Sounds good, we probably also want to actually do a round of releases as well before merging and releasing this.

**Comment by simolus3** [comment]
> Sounds good, we probably also want to actually do a round of releases as well before merging and releasing this.

Definitely. And if we end up doing a `build: 2.0.0` release for this, we might want to integrate [other breaking changes](https://github.com/dart-lang/build/labels/next-breaking-release) too.

**Comment by jakemac53** [comment]
@simolus3 we should be ready to proceed here if you want to merge with master and I can take another pass

**Comment by jakemac53** [comment]
Note that our tests red on master right now, I am working on fixing them.

**Comment by jakemac53** [comment]
Re-running the tests now, should go green

**Comment by jakemac53** [comment]
@simolus3 I think maybe you need to merge with master again to get the test fixes?

File: _test/pubspec.yaml
Comment (jakemac53) [inline _test/pubspec.yaml:20]: This shouldn't be necessary any more (we should be on `1.x.x` everywhere now)

File: build_test/mono_pkg.yaml
Comment (jakemac53) [inline build_test/mono_pkg.yaml:10]: We shouldn't need this if the tests themselves are opted out (this should be the default for opted out tests).

File: build_test/CHANGELOG.md
Comment (simolus3) [inline build_test/CHANGELOG.md:3]: Just for completeness, I want to point out that this package is not fully migrated yet (and shouldn't be released as-is). We're still blocked by `build_resolvers`, which is blocked by `graphs`.

Comment (jakemac53) [reply build_test/CHANGELOG.md:3]: Understood, did you plan to look at those or should one of us (either is fine, just want to not duplicate work :) )

**Review by jakemac53 — APPROVED** [review]

File: build_test/CHANGELOG.md
Comment (simolus3) [reply build_test/CHANGELOG.md:3]: One of you would have to take a look at https://github.com/dart-lang/graphs/pull/52. As soon as that package is migrated I'm happy to work on `build_resolvers` and the rest of `build_test`.

**Comment by NarHakobyan** [comment]
are you planning to release?

**Comment by jakemac53** [comment]
Probably sometime next week, there is another small breaking change we are rolling as a part of this release (https://github.com/dart-lang/build/issues/1899) but its going to take a little bit to roll that out (and I am OOO for the rest of this week).

**Pull request merged**
